six/modules/c++/cphd/source/PVPArray.cpp

```diff
@@ -0,0 +1,844 @@
+/* =========================================================================
+ * This file is part of cphd-c++
+ * =========================================================================
+ *
+ * (C) Copyright 2004 - 2019, MDA Information Systems LLC
+ *
+ * cphd-c++ is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <ostream>
+#include <vector>
+#include <stddef.h>
+#include <typeinfo>
+
+#include <six/Init.h>
+#include <sys/Conf.h>
+#include <cphd/Types.h>
+#include <cphd/PVPArray.h>
+#include <cphd/Metadata.h>
+
+namespace
+{
+// Set data from data block into data struct
+template <typename T> inline void setData(const sys::byte* data,
+                    T& dest)
+{
+    memcpy(&dest, data, sizeof(T));
+}
+
+inline void setData(const sys::byte* data,
+                    cphd::Vector3& dest)
+{
+    setData(data, dest[0]);
+    setData(data, dest[1]);
+    setData(data, dest[2]);
+}
+
+// Get data from data struct and put into data block
+template <typename T> inline void getData(sys::ubyte* dest,
+                    T value)
+{
+    memcpy(dest, &value, sizeof(T));
+}
+
+inline void getData(sys::ubyte* dest,
+                    const cphd::Vector3& value)
+{
+    getData(dest, value[0]);
+    getData(dest, value[1]);
+    getData(dest, value[2]);
+}
+}
+
+namespace cphd
+{
+
+PVPArray::PVPSet::PVPSet() :
+    txTime(six::Init::undefined<double>()),
+    txPos(six::Init::undefined<Vector3>()),
+    txVel(six::Init::undefined<Vector3>()),
+    rcvTime(six::Init::undefined<double>()),
+    rcvPos(six::Init::undefined<Vector3>()),
+    rcvVel(six::Init::undefined<Vector3>()),
+    srpPos(six::Init::undefined<Vector3>()),
+    aFDOP(six::Init::undefined<double>()),
+    aFRR1(six::Init::undefined<double>()),
+    aFRR2(six::Init::undefined<double>()),
+    fx1(six::Init::undefined<double>()),
+    fx2(six::Init::undefined<double>()),
+    toa1(six::Init::undefined<double>()),
+    toa2(six::Init::undefined<double>()),
+    tdTropoSRP(six::Init::undefined<double>()),
+    sc0(six::Init::undefined<double>()),
+    scss(six::Init::undefined<double>())
+{
+}
+
+void PVPArray::PVPSet::write(Pvp& p, const sys::byte* input)
+{
+    ::setData(input + p.txPos.getOffset()*8, txTime);
```

Comment:
How about adding a `getByteOffset()` so you don't have to multiply by 8 all the time?